File: sdk/runanywhere-react-native/src/components/SpeakerDiarization/SpeakerDiarizationComponent.ts
/**
 * SpeakerDiarizationComponent.ts
 *
 * Speaker Diarization component following the clean architecture
 *
 * Reference: sdk/runanywhere-swift/Sources/RunAnywhere/Components/SpeakerDiarization/SpeakerDiarizationComponent.swift
 */

import { BaseComponent } from '../../Core/Components/BaseComponent';
import { SDKComponent } from '../../Core/Models/Common/SDKComponent';
import { ModuleRegistry } from '../../Core/ModuleRegistry';
import { SDKError, SDKErrorCode } from '../../Public/Errors/SDKError';
import type { SpeakerDiarizationConfiguration } from './SpeakerDiarizationConfiguration';
import type {
  SpeakerDiarizationInput,
  SpeakerDiarizationOutput,
  SpeakerSegment,
  SpeakerProfile,
  LabeledTranscription,
  SpeakerInfo,
} from './SpeakerDiarizationModels';
import type { SpeakerDiarizationService } from '../../Core/Protocols/Voice/SpeakerDiarizationService';
import type { SpeakerDiarizationServiceProvider } from '../../Core/Protocols/Voice/SpeakerDiarizationServiceProvider';
import { AnyServiceWrapper } from '../../Core/Components/BaseComponent';

/**
 * Default Speaker Diarization Service implementation
 */
class DefaultSpeakerDiarizationService implements SpeakerDiarizationService {
  private speakers: Map<string, SpeakerInfo> = new Map();
  private speakerCounter = 0;

  async initialize(): Promise<void> {
    // No initialization needed for default implementation
  }

  async processAudio(
    audioData: string | ArrayBuffer,
    sampleRate?: number
  ): Promise<import('../../Core/Models/SpeakerDiarization/SpeakerDiarizationResult').SpeakerDiarizationResult> {
    // Simple energy-based speaker detection
    // In real implementation, this would use ML models
    const speakerId = `speaker_${this.speakerCounter++}`;
    const speakerInfo: SpeakerInfo = {
      id: speakerId,
      name: null,
      confidence: 0.8,
      embedding: null,
    };
    this.speakers.set(speakerId, speakerInfo);

    // Return result matching SpeakerDiarizationResult interface
    return {
      segments: [],
      speakers: Array.from(this.speakers.values()).map((s) => ({
        id: s.id,
        name: s.name,
        confidence: s.confidence,
        embedding: s.embedding,
      })),
    };
  }

  get isReady(): boolean {
    return true;
  }

  async cleanup(): Promise<void> {
    this.speakers.clear();
    this.speakerCounter = 0;
  }
}

/**
 * Speaker Diarization Service Wrapper
 */
export class SpeakerDiarizationServiceWrapper extends AnyServiceWrapper<SpeakerDiarizationService> {
  constructor(service: SpeakerDiarizationService | null = null) {
    super(service);
  }
}

/**
 * Speaker Diarization component
 */
export class SpeakerDiarizationComponent extends BaseComponent<SpeakerDiarizationServiceWrapper> {
  // MARK: - Properties

  public static override componentType: SDKComponent = SDKComponent.SpeakerDiarization;

  private readonly diarizationConfiguration: SpeakerDiarizationConfiguration;
  private speakerProfiles: Map<string, SpeakerProfile> = new Map();
  private isServiceReady = false;

  // MARK: - Initialization

  constructor(configuration: SpeakerDiarizationConfiguration) {
    super(configuration);
    this.diarizationConfiguration = configuration;
  }

  // MARK: - Service Creation

  protected override async createService(): Promise<SpeakerDiarizationServiceWrapper> {
    // Try to get a registered speaker diarization provider from central registry
    const provider = ModuleRegistry.shared.speakerDiarizationProvider();

    if (provider) {
      try {
        const diarizationService = await provider.createSpeakerDiarizationService(
          this.diarizationConfiguration
        );
        return new SpeakerDiarizationServiceWrapper(diarizationService);
      } catch (error) {
        // Fall through to default
      }
    }

    // Fallback to default implementation
    const defaultService = new DefaultSpeakerDiarizationService();
    await defaultService.initialize();
    return new SpeakerDiarizationServiceWrapper(defaultService);
  }

  protected override async initializeService(): Promise<void> {
    if (this.service?.wrappedService) {
      await this.service.wrappedService.initialize();
    }
    this.isServiceReady = true;
  }

  protected override async performCleanup(): Promise<void> {
    if (this.service?.wrappedService) {
      await this.service.wrappedService.cleanup();
    }
    this.speakerProfiles.clear();
    this.isServiceReady = false;
  }

  // MARK: - Public API

  /**
   * Diarize audio to identify speakers
   */
  public async diarize(
    audioData: Buffer | Uint8Array,
    format: string = 'wav'
  ): Promise<SpeakerDiarizationOutput> {
    this.ensureReady();

    const input: SpeakerDiarizationInput = {
      audioData,
      format,
      transcription: null,
      expectedSpeakers: null,
      options: null,
      validate: () => {
        if (!audioData || audioData.length === 0) {
          throw new SDKError(SDKErrorCode.ValidationFailed, 'Audio data cannot be empty');
        }
      },
      timestamp: new Date(),
    };

    return await this.process(input);
  }

  /**
   * Diarize with transcription for labeled output
   */
  public async diarizeWithTranscription(
    audioData: Buffer | Uint8Array,
    transcription: SpeakerDiarizationInput['transcription'],
    format: string = 'wav'
  ): Promise<SpeakerDiarizationOutput> {
    this.ensureReady();

    const input: SpeakerDiarizationInput = {
      audioData,
      format,
      transcription,
      expectedSpeakers: null,
      options: null,
      validate: () => {
        if (!audioData || audioData.length === 0) {
          throw new SDKError(SDKErrorCode.ValidationFailed, 'Audio data cannot be empty');
        }
      },
      timestamp: new Date(),
    };

    return await this.process(input);
  }

  /**
   * Process diarization input
   */
  public async process(input: SpeakerDiarizationInput): Promise<SpeakerDiarizationOutput> {
    this.ensureReady();

    if (!this.service?.wrappedService) {
      throw new SDKError(SDKErrorCode.ComponentNotReady, 'Speaker diarization service not available');
    }

    // Validate input
    input.validate();

    // Track processing time
    const startTime = Date.now();

    // Convert audio data to ArrayBuffer for service
    const audioBuffer = Buffer.isBuffer(input.audioData)
      ? input.audioData.buffer
      : input.audioData instanceof Uint8Array
      ? input.audioData.buffer
      : Buffer.from(input.audioData).buffer;

    // Process audio to detect speakers
    const result = await this.service.wrappedService.processAudio(audioBuffer, 16000);

    // Build segments from result
    const segments: SpeakerSegment[] = result.segments?.map((seg) => ({
      speakerId: seg.speakerId ?? 'unknown',
      startTime: seg.startTime ?? 0,
      endTime: seg.endTime ?? 0,
      confidence: seg.confidence ?? 0.8,
    })) ?? [];

    // Build speaker profiles from result
    const allSpeakers = result.speakers ?? [];
    const profiles: SpeakerProfile[] = allSpeakers.map((speaker) => {
      const speakerId = speaker.id ?? 'unknown';
      const speakerSegments = segments.filter((s) => s.speakerId === speakerId);
      const totalTime = speakerSegments.reduce((sum, seg) => sum + (seg.endTime - seg.startTime), 0);
      return {
        id: speakerId,
        embedding: speaker.embedding ?? null,
        totalSpeakingTime: totalTime,
        segmentCount: speakerSegments.length,
        name: speaker.name ?? null,
      };
    });

    // Store profiles
    profiles.forEach((profile) => {
      this.speakerProfiles.set(profile.id, profile);
    });

    // Create labeled transcription if provided
    let labeledTranscription: LabeledTranscription | null = null;
    if (input.transcription && input.transcription.wordTimestamps) {
      labeledTranscription = this.createLabeledTranscription(
        input.transcription.wordTimestamps,
        segments
      );
    }

    const processingTime = (Date.now() - startTime) / 1000; // seconds

    // Calculate audio length from segments
    const audioLength = segments.length > 0
      ? Math.max(...segments.map((s) => s.endTime))
      : 0;

    const metadata = {
      processingTime,
      audioLength,
      speakerCount: profiles.length,
      method: this.diarizationConfiguration.modelId ? 'ml' : 'energy',
    };

    return {
      segments,
      speakers: profiles,
      labeledTranscription,
      metadata,
      timestamp: new Date(),
    };
  }

  // MARK: - Helper Methods

  /**
   * Convert audio data to float array
   */
  private convertDataToFloatArray(data: Buffer | Uint8Array): number[] {
    // Convert to Float32 array
    const buffer = Buffer.isBuffer(data) ? data : Buffer.from(data);
    const floatArray: number[] = [];
    for (let i = 0; i < buffer.length; i += 4) {
      if (i + 3 < buffer.length) {
        floatArray.push(buffer.readFloatLE(i));
      }
    }
    return floatArray;
  }

  /**
   * Create labeled transcription from word timestamps and segments
   */
  private createLabeledTranscription(
    wordTimestamps: Array<{ word: string; startTime: number; endTime: number; confidence: number }>,
    segments: SpeakerSegment[]
  ): LabeledTranscription {
    const labeledSegments = wordTimestamps.map((word) => {
      // Find which speaker segment this word belongs to
      const segment = segments.find(
        (s) => word.startTime >= s.startTime && word.endTime <= s.endTime
      );
      return {
        speakerId: segment?.speakerId ?? 'unknown',
        text: word.word,
        startTime: word.startTime,
        endTime: word.endTime,
      };
    });

    return { segments: labeledSegments };
  }
}

File: sdk/runanywhere-kotlin/src/commonMain/kotlin/com/runanywhere/sdk/features/speakerdiarization/SpeakerDiarizationComponent.kt
package com.runanywhere.sdk.features.speakerdiarization

import com.runanywhere.sdk.core.capabilities.*
import com.runanywhere.sdk.data.models.SDKError
import com.runanywhere.sdk.events.EventBus
import com.runanywhere.sdk.foundation.ServiceContainer
import com.runanywhere.sdk.utils.getCurrentTimeMillis
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock

/**
 * Speaker Diarization Component
 * Matches iOS SpeakerDiarizationComponent architecture exactly
 */
class SpeakerDiarizationComponent(
    configuration: SpeakerDiarizationConfiguration,
    serviceContainer: ServiceContainer? = null,
) : BaseComponent<SpeakerDiarizationService>(configuration, serviceContainer) {
    // MARK: - Component Properties

    override val componentType: SDKComponent = SDKComponent.SPEAKER_DIARIZATION

    private val diarizationConfiguration: SpeakerDiarizationConfiguration
        get() = configuration as SpeakerDiarizationConfiguration

    private val processingMutex = Mutex()
    private var isProcessing = false

    // Event callbacks matching iOS implementation
    private var onSpeakerDetected: ((SpeakerInfo) -> Unit)? = null
    private var onSpeakerChanged: ((previous: SpeakerInfo?, current: SpeakerInfo) -> Unit)? = null
    private var onSegmentCompleted: ((SpeakerSegment) -> Unit)? = null
    private var onProcessingCompleted: ((SpeakerDiarizationOutput) -> Unit)? = null
    private var onError: ((SpeakerDiarizationError) -> Unit)? = null

    // MARK: - Service Creation

    override suspend fun createService(): SpeakerDiarizationService {
        // Create the default speaker diarization service directly
        // Simplified implementation without provider pattern
        val database = InMemorySpeakerDatabase()
        val audioProcessor = PlatformAudioProcessor()
        return DefaultSpeakerDiarizationService(diarizationConfiguration, database, audioProcessor)
    }

    override suspend fun initializeService() {
        service?.initialize(diarizationConfiguration.modelId)
            ?: throw SDKError.ServiceNotAvailable("Speaker Diarization service not available")
    }

    // MARK: - Processing Methods

    /**
     * Process audio for real-time speaker detection
     */
    suspend fun processAudio(input: SpeakerDiarizationInput): SpeakerDiarizationOutput {
        ensureReady()

        return processingMutex.withLock {
            if (isProcessing && !diarizationConfiguration.enableRealTimeProcessing) {
                throw SDKError.InvalidState("Component is already processing audio")
            }

            isProcessing = true
            val startTime = System.currentTimeMillis()

            try {
                val service =
                    this.service
                        ?: throw SDKError.ServiceNotAvailable("Speaker Diarization service not available")

                input.validate()
                state = ComponentState.PROCESSING

                // Emit processing started event
                publishEvent(SpeakerDiarizationEvent.ProcessingStarted)

                val result: SpeakerDiarizationOutput

                if (diarizationConfiguration.batchProcessingMode && input.audioBuffer != null) {
                    // Batch processing for complete audio
                    result = performBatchDiarization(input, service)
                } else {
                    // Real-time processing
                    result = performRealTimeDiarization(input, service)
                }

                // Emit processing completed event
                publishEvent(SpeakerDiarizationEvent.ProcessingCompleted)
                onProcessingCompleted?.invoke(result)

                state = ComponentState.READY
                return result
            } catch (e: Exception) {
                state = ComponentState.FAILED
                val error =
                    when (e) {
                        is SpeakerDiarizationError -> e
                        else -> SpeakerDiarizationError.ProcessingFailed(e.message ?: "Unknown processing error")
                    }
                publishEvent(SpeakerDiarizationEvent.Error(error))
                onError?.invoke(error)
                throw error
            } finally {
                isProcessing = false
            }
        }
    }

    /**
     * Stream audio processing for real-time diarization
     */
    fun processAudioStream(audioFlow: Flow<ByteArray>): Flow<SpeakerInfo> =
        flow {
            ensureReady()

            val service =
                this@SpeakerDiarizationComponent.service
                    ?: throw SDKError.ServiceNotAvailable("Speaker Diarization service not available")

            audioFlow.collect { audioData ->
                try {
                    // Convert audio data to float array (assuming 16-bit PCM)
                    val samples = convertBytesToFloatArray(audioData)

                    // Process audio chunk
                    val speaker = service.processAudio(samples)

                    // Emit speaker detection events
                    if (speaker.id != "SILENCE") {
                        onSpeakerDetected?.invoke(speaker)
                        publishEvent(SpeakerDiarizationEvent.SpeakerDetected(speaker))
                    }

                    emit(speaker)
                } catch (e: Exception) {
                    val error = SpeakerDiarizationError.ProcessingFailed(e.message ?: "Stream processing error")
                    publishEvent(SpeakerDiarizationEvent.Error(error))
                    throw error
                }
            }
        }

    /**
     * Get all detected speakers
     */
    fun getAllSpeakers(): List<SpeakerInfo> {
        ensureReady()
        return service?.getAllSpeakers() ?: emptyList()
    }

    /**
     * Get speaker profile by ID
     */
    suspend fun getSpeakerProfile(speakerId: String): SpeakerProfile? {
        ensureReady()
        return service?.getSpeakerProfile(speakerId)
    }

    /**
     * Update speaker name/label
     */
    suspend fun updateSpeakerName(
        speakerId: String,
        name: String,
    ) {
        ensureReady()
        service?.updateSpeakerName(speakerId, name)
    }

    /**
     * Reset all speaker data
     */
    suspend fun reset() {
        service?.reset()
    }

    /**
     * Reset speaker profiles only
     */
    suspend fun resetProfiles() {
        service?.resetProfiles()
    }

    // MARK: - Event Handling

    /**
     * Set callback for speaker detection
     */
    fun onSpeakerDetected(callback: (SpeakerInfo) -> Unit) {
        this.onSpeakerDetected = callback
    }

    /**
     * Set callback for speaker change
     */
    fun onSpeakerChanged(callback: (previous: SpeakerInfo?, current: SpeakerInfo) -> Unit) {
        this.onSpeakerChanged = callback
    }

    /**
     * Set callback for segment completion
     */
    fun onSegmentCompleted(callback: (SpeakerSegment) -> Unit) {
        this.onSegmentCompleted = callback
    }

    /**
     * Set callback for processing completion
     */
    fun onProcessingCompleted(callback: (SpeakerDiarizationOutput) -> Unit) {
        this.onProcessingCompleted = callback
    }

    /**
     * Set callback for errors
     */
    fun onError(callback: (SpeakerDiarizationError) -> Unit) {
        this.onError = callback
    }

    // MARK: - Private Processing Methods

    /**
     * Perform batch diarization on complete audio
     */
    private suspend fun performBatchDiarization(
        input: SpeakerDiarizationInput,
        service: SpeakerDiarizationService,
    ): SpeakerDiarizationOutput {
        val audioBuffer =
            input.audioBuffer
                ?: throw SDKError.ValidationFailed("Audio buffer required for batch processing")

        val result =
            service.performDetailedDiarization(audioBuffer, input.sampleRate)
                ?: throw SpeakerDiarizationError.ProcessingFailed("Diarization returned no results")

        // Create speaker profiles from results
        val speakerProfiles =
            result.speakers.map { speaker ->
                SpeakerProfile(
                    id = speaker.id,
                    embedding = speaker.embedding,
                    totalSpeakingTime =
                        result.segments
                            .filter { it.speakerId == speaker.id }
                            .sumOf { it.speechDuration },
                    segmentCount = result.segments.count { it.speakerId == speaker.id },
                    name = speaker.name,
                    averageConfidence = result.confidence,
                    lastUpdated = getCurrentTimeMillis(),
                )
            }

        // Create labeled transcription if STT output provided
        val labeledTranscription =
            input.transcription?.let { sttOutput ->
                if (sttOutput.wordTimestamps != null) {
                    TranscriptionSpeakerIntegrator.createLabeledTranscription(
                        sttOutput.wordTimestamps,
                        result.segments,
                        speakerProfiles,
                    )
                } else {
                    TranscriptionSpeakerIntegrator.mapTranscriptionToSpeakers(
                        sttOutput.text,
                        result.segments,
                        speakerProfiles,
                    )
                }
            }

        // Create metadata
        val metadata =
            DiarizationMetadata(
                processingTime = result.processingTime,
                audioLength = audioBuffer.size.toDouble() / input.sampleRate,
                speakerCount = result.speakers.size,
                method = "energy",
                averageConfidence = result.confidence,
                segmentCount = result.segments.size,
            )

        return SpeakerDiarizationOutput(
            segments = result.segments,
            speakers = speakerProfiles,
            labeledTranscription = labeledTranscription,
            metadata = metadata,
        )
    }

    /**
     * Perform real-time diarization on audio chunk
     */
    private suspend fun performRealTimeDiarization(
        input: SpeakerDiarizationInput,
        service: SpeakerDiarizationService,
    ): SpeakerDiarizationOutput {
        val samples =
            if (input.audioBuffer != null) {
                input.audioBuffer
            } else {
                convertBytesToFloatArray(input.audioData)
            }

        // Process audio chunk
        val detectedSpeaker = service.processAudio(samples)

        // Create simple segment for real-time processing
        val segment =
            SpeakerSegment(
                speakerId = detectedSpeaker.id,
                startTime = 0.0,
                endTime = samples.size.toDouble() / input.sampleRate,
                confidence = detectedSpeaker.confidence ?: 0.8f,
            )

        // Get all speakers for profiles
        val allSpeakers = service.getAllSpeakers()
        val speakerProfiles =
            allSpeakers.map { speaker ->
                SpeakerProfile(
                    id = speaker.id,
                    embedding = speaker.embedding,
                    name = speaker.name,
                    lastUpdated = getCurrentTimeMillis(),
                )
            }

        // Create metadata for real-time processing
        val audioLength = samples.size.toDouble() / input.sampleRate
        val metadata =
            DiarizationMetadata(
                processingTime = 0.05, // Estimated for real-time
                audioLength = audioLength,
                speakerCount = allSpeakers.size,
                method = "energy",
                averageConfidence = detectedSpeaker.confidence ?: 0.8f,
                segmentCount = 1,
            )

        return SpeakerDiarizationOutput(
            segments = listOf(segment),
            speakers = speakerProfiles,
            labeledTranscription = null, // Real-time processing doesn't include transcription
            metadata = metadata,
        )
    }

    // MARK: - Utility Methods

    /**
     * Convert byte array to float array (assuming 16-bit PCM)
     */
    private fun convertBytesToFloatArray(audioData: ByteArray): FloatArray {
        val samples = FloatArray(audioData.size / 2)
        for (i in samples.indices) {
            val index = i * 2
            val sample =
                (audioData[index].toInt() and 0xFF) or
                    (audioData[index + 1].toInt() shl 8)
            samples[i] = sample.toShort() / 32768.0f
        }
        return samples
    }

    /**
     * Publish event to event bus
     */
    private fun publishEvent(event: SpeakerDiarizationEvent) {
        try {
            EventBus.publish(event)
        } catch (e: Exception) {
            // Ignore event publishing errors to avoid affecting main processing
        }
    }

    // MARK: - Cleanup

    override suspend fun performCleanup() {
        super.performCleanup()

        // Clear callbacks
        onSpeakerDetected = null
        onSpeakerChanged = null
        onSegmentCompleted = null
        onProcessingCompleted = null
        onError = null

        // Reset processing state
        isProcessing = false
    }
}
